satellite/gracefulexit/endpoint.go

```diff
@@ -234,40 +234,40 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 			close(processChan)
 		}()
 
-		for range ticker.C {
-			// exit if context canceled
+	loop:
+		for {
 			select {
 			case <-ctx.Done():
+				// exit if context canceled
 				return ctx.Err()
-			default:
-			}
-
-			if pending.length() == 0 {
-				incomplete, err := endpoint.db.GetIncompleteNotFailed(ctx, nodeID, endpoint.config.EndpointBatchSize, 0)
-				if err != nil {
-					return handleError(err)
-				}
-
-				if len(incomplete) == 0 {
-					incomplete, err = endpoint.db.GetIncompleteFailed(ctx, nodeID, endpoint.config.MaxFailuresPerPiece, endpoint.config.EndpointBatchSize, 0)
+			case <-ticker.C:
```

Comment:
I think we still need to call ticker.Stop otherwise we would leak a channel?

Reply:
https://golang.org/pkg/time/#Ticker

> Stop the ticker to release associated resources.

So, yes, you need to always stop it to prevent leaking memory.

Reply:
Also, is there a reason it's not using a cycle?

Reply:
I don't think there is any reason to not use a cycle. I will change it.

Reply:
@egonelbre updated

Reply:
@egonelbre updated